File: .config/nvim/testing/sample.py
#!/usr/bin/env python3

# Sample Python file with various issues for testing

import os
import sys
import json
from typing import List, Dict, Optional, Any
from datetime import datetime

# Unused import
import time
import random  # Another unused import
import subprocess  # Unused import

# Constants with potential issues
API_URL = "https://api.example.com"
MAX_RETRIES = 3
TIMEOUT = 5.0
UNUSED_CONSTANT = "This is not used"
ANOTHER_UNUSED = 42

# Global variable (should be in function)
global_var = "This pollutes global scope"

# Class with potential issues
class Person:
    def __init__(self, name: str, age: int, email: str = ""):
        self.name = name
        self.age = age
        self.email = email
        self.created_at = datetime.now()

    def greet(self, greeting: str = "Hello") -> str:
        return f"{greeting}, {self.name}!"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "age": self.age,
            "email": self.email,
            "created_at": self.created_at.isoformat()
        }

    def __str__(self) -> str:
        return f"Person(name='{self.name}', age={self.age})"

# Function with potential issues
def create_people(names: List[str]) -> List[Person]:
    people = []
    for i, name in enumerate(names):
        age = 20 + i * 5  # Potential issue: hardcoded logic
        person = Person(name, age)
        people.append(person)

    # Unused variable
    unused_var = "This is not used"
    another_unused = 123

    # Unused parameter in nested function
    def helper_func(param1, param2, param3):
        return param1

    return people

def process_data(data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    processed = []
    for item in data:
        processed_item = {
            "id": item.get("id"),
            "name": item.get("name", "").upper(),
            "processed": True,
            "timestamp": datetime.now().isoformat()
        }
        processed.append(processed_item)

    return processed

# Async function with potential issues
async def fetch_user_data(user_id: int) -> Optional[Dict[str, Any]]:
    try:
        # Simulated async operation
        await asyncio.sleep(0.1)  # Missing import for asyncio
        return {
            "id": user_id,
            "name": f"User {user_id}",
            "email": f"user{user_id}@example.com",
            "active": True
        }
    except Exception as e:
        print(f"Error fetching user {user_id}: {e}")
        return None
        # Missing rethrow or proper error handling

# Context manager with potential issues
class FileHandler:
    def __init__(self, filename: str, mode: str = "r"):
        self.filename = filename
        self.mode = mode
        self.file = None

    def __enter__(self):
        self.file = open(self.filename, self.mode)
        return self.file

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.file:
            self.file.close()

# Decorator with potential issues
def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()
        print(f"{func.__name__} took {end - start:.4f} seconds")
        return result
    return wrapper

@timing_decorator
def expensive_operation(n: int) -> int:
    total = 0
    for i in range(n):
        total += i * i
    return total

# Main function with potential issues
def main():
    # Create people
    names = ["Alice", "Bob", "Charlie"]
    people = create_people(names)

    for person in people:
        print(person.greet())
        print(person.to_dict())

    # Process data
    sample_data = [
        {"id": 1, "name": "alice"},
        {"id": 2, "name": "bob"},
        {"id": 3, "name": "charlie"}
    ]

    processed = process_data(sample_data)
    print("Processed data:", json.dumps(processed, indent=2))

    # File operations
    with FileHandler("output.txt", "w") as f:
        f.write("Sample output\n")
        f.write(f"Generated at: {datetime.now()}\n")

    # Expensive operation
    result = expensive_operation(1000)
    print(f"Expensive operation result: {result}")

    # Type checking with potential issues
    value: Any = "test"
    if isinstance(value, str):
        print(f"String value: {value.upper()}")

    # List comprehension with potential issues
    squares = [x * x for x in range(10) if x % 2 == 0]
    print(f"Squares of even numbers: {squares}")

    # Unused variable in main
    main_unused = "This is not used"

    # Missing return statement
    # return None

if __name__ == "__main__":
    main()

File: .config/nvim/testing/sample.rs
// Sample Rust file with various issues for testing

use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Write};
use std::time::{SystemTime, UNIX_EPOCH};

// Unused import
use std::path::Path;
use std::collections::HashSet; // Another unused import

// Constants with potential issues
const API_URL: &str = "https://api.example.com";
const MAX_RETRIES: u32 = 3;
const TIMEOUT: u64 = 5000;
const UNUSED_CONSTANT: &str = "This is not used";
const ANOTHER_UNUSED: i32 = 42;

// Struct with potential issues
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: u32,
    email: Option<String>,
    created_at: u64,
}

impl Person {
    fn new(name: String, age: u32, email: Option<String>) -> Self {
        Self {
            name,
            age,
            email,
            created_at: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        }
    }

    fn greet(&self, greeting: Option<&str>) -> String {
        let greeting = greeting.unwrap_or("Hello");
        format!("{}, {}!", greeting, self.name)
    }

    fn to_dict(&self) -> HashMap<String, String> {
        let mut map = HashMap::new();
        map.insert("name".to_string(), self.name.clone());
        map.insert("age".to_string(), self.age.to_string());
        map.insert("created_at".to_string(), self.created_at.to_string());

        if let Some(email) = &self.email {
            map.insert("email".to_string(), email.clone());
        }

        map
    }
}

// Enum with potential issues
#[derive(Debug, PartialEq)]
enum Status {
    Active,
    Inactive,
    Pending,
    Suspended,
}

// Trait with potential issues
trait Processable {
    fn process(&self) -> Result<String, String>;
    fn validate(&self) -> bool;
}

impl Processable for Person {
    fn process(&self) -> Result<String, String> {
        if self.age < 18 {
            return Err("Person is underage".to_string());
        }
        Ok(format!("Processed: {}", self.name))
    }

    fn validate(&self) -> bool {
        !self.name.is_empty() && self.age > 0 && self.age <= 120
    }
}

// Function with potential issues
fn create_people(names: Vec<String>) -> Vec<Person> {
    let mut people = Vec::new();

    for (i, name) in names.iter().enumerate() {
        let age = (20 + i as u32).min(100); // Potential issue: arbitrary age calculation
        let person = Person::new(name.clone(), age, None);
        people.push(person);
    }

    // Unused variable
    let unused_var = "This is not used";
    let another_unused = 123;

    // Unused parameter in nested function
    let unused_param = |param1: i32, param2: String, param3: bool| param1;

    people
}

// Generic function with potential issues
fn wrap_in_option<T>(value: T) -> Option<T> {
    Some(value)
}

// Result handling with potential issues
fn divide(a: f64, b: f64) -> Result<f64, String> {
    if b == 0.0 {
        Err("Division by zero".to_string())
    } else {
        Ok(a / b)
    }
}

// Async function (requires async runtime)
async fn fetch_user_data(user_id: u32) -> Result<HashMap<String, String>, String> {
    // Simulated async operation
    tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;

    let mut user_data = HashMap::new();
    user_data.insert("id".to_string(), user_id.to_string());
    user_data.insert("name".to_string(), format!("User {}", user_id));
    user_data.insert("email".to_string(), format!("user{}@example.com", user_id));
    user_data.insert("active".to_string(), "true".to_string());

    Ok(user_data)
}

// Function that's never called
fn never_called() {
    println!("This function is never called");
}

// Function with panic (should use Result)
fn bad_function(x: i32) -> String {
    if x < 0 {
        panic!("This should not panic!"); // Should return Result instead
    }
    format!("Value: {}", x)
}

// File operations with potential issues
fn write_sample_file() -> io::Result<()> {
    let mut file = File::create("sample.txt")?;

    writeln!(file, "Sample Rust output")?;
    writeln!(file, "Generated at: {:?}", SystemTime::now())?;

    // Missing explicit flush
    // file.flush()?;

    Ok(())
}

// Error handling with custom error type
#[derive(Debug)]
enum AppError {
    IoError(io::Error),
    ParseError(String),
    ValidationError(String),
}

impl From<io::Error> for AppError {
    fn from(error: io::Error) -> Self {
        AppError::IoError(error)
    }
}

impl std::fmt::Display for AppError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AppError::IoError(e) => write!(f, "IO Error: {}", e),
            AppError::ParseError(msg) => write!(f, "Parse Error: {}", msg),
            AppError::ValidationError(msg) => write!(f, "Validation Error: {}", msg),
        }
    }
}

// Function returning custom error
fn validate_person(person: &Person) -> Result<(), AppError> {
    if !person.validate() {
        return Err(AppError::ValidationError(
            "Invalid person data".to_string()
        ));
    }
    Ok(())
}

// Main function with potential issues
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create people
    let names = vec![
        "Alice".to_string(),
        "Bob".to_string(),
        "Charlie".to_string(),
    ];

    let people = create_people(names);

    for person in &people {
        println!("{}", person.greet(None));

        match person.process() {
            Ok(result) => println!("Process result: {}", result),
            Err(e) => println!("Process error: {}", e),
        }

        // Validate person
        if let Err(e) = validate_person(person) {
            println!("Validation error: {}", e);
        }
    }

    // Test division
    match divide(10.0, 2.0) {
        Ok(result) => println!("Division result: {}", result),
        Err(e) => println!("Division error: {}", e),
    }

    match divide(10.0, 0.0) {
        Ok(result) => println!("Division result: {}", result),
        Err(e) => println!("Division error: {}", e),
    }

    // Test generic function
    let wrapped_name = wrap_in_option("Test Name");
    println!("Wrapped name: {:?}", wrapped_name);

    // Write file
    write_sample_file()?;

    // HashMap operations
    let mut config = HashMap::new();
    config.insert("api_url".to_string(), API_URL.to_string());
    config.insert("timeout".to_string(), TIMEOUT.to_string());
    config.insert("retries".to_string(), MAX_RETRIES.to_string());

    for (key, value) in &config {
        println!("Config {}: {}", key, value);
    }

    // Vector operations
    let numbers: Vec<i32> = (1..=10).collect();
    let squares: Vec<i32> = numbers.iter().map(|x| x * x).collect();
    let even_squares: Vec<i32> = squares.iter().filter(|x| x % 2 == 0).copied().collect();

    println!("Even squares: {:?}", even_squares);

    // Unused variable in main
    let main_unused = "This is not used";

    // Complex expression that should be split
    let result = if people.len() > 0 && people[0].age > 18 && people[0].email.is_some() {
        format!("First person is valid: {}", people[0].name)
    } else {
        "First person is invalid".to_string()
    };
    println!("{}", result);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_person_creation() {
        let person = Person::new("Test".to_string(), 25, Some("test@example.com".to_string()));
        assert_eq!(person.name, "Test");
        assert_eq!(person.age, 25);
    }

    #[test]
    fn test_divide() {
        assert_eq!(divide(10.0, 2.0).unwrap(), 5.0);
        assert!(divide(10.0, 0.0).is_err());
    }
}
